pyqtorch/apply.py

```diff
@@ -57,3 +60,63 @@ def apply_operator(
         map(lambda e: "".join(list(e)), [operator_dims, in_state_dims, out_state_dims])
     )
     return einsum(f"{operator_dims},{in_state_dims}->{out_state_dims}", operator, state)
+
+
+def apply_ope_ope(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
+    """
+    Compute the product of two operators.
+
+    Args:
+        operator_1 (Tensor): The first operator.
+        operator_2 (Tensor): The second operator.
+        target (int): The target qubit index.
+
+    Returns:
+        Tensor: The product of the two operators.
+    """
+
+    # Dimension verifications:
+    n_qubits_1 = int(log2(operator_1.size(1)))
+    n_qubits_2 = int(log2(operator_2.size(1)))
+    batch_size_1 = operator_1.size(-1)
+    batch_size_2 = operator_2.size(-1)
+    if n_qubits_1 != n_qubits_2:
+        if n_qubits_1 > n_qubits_2:
+            operator_2 = promote_ope(operator_2, target, n_qubits_1)
+        if n_qubits_1 < n_qubits_2:
+            operator_1 = promote_ope(operator_1, target, n_qubits_2)
+    if batch_size_1 != batch_size_2:
+        if batch_size_1 > batch_size_2:
+            operator_2 = operator_2.repeat(1, 1, batch_size_1)
+        if batch_size_2 > batch_size_1:
+            operator_1 = operator_1.repeat(1, 1, batch_size_2)
+
+    # Permute the batch size on first dimension to allow torch.bmm():
+    def batch_first(operator: Tensor) -> Tensor:
```

Comment:
Would it make sense to add this function to `utils`? I see it reused many times throughout PyQ. It seems like a bit of a waste to redefine it every time.

Reply:
For now, I'm using this function exclusively here, but indeed, it could be a good idea to provide access to this function more generally for future use or for user manipulation. I'll move it to `pyqtorch/utils.py` .

Reply:
Done.